### PR TITLE
sorting an empty view should exit early and not fail

### PR DESCRIPTION
**Pull request by fnrizzi**
## `sort`
The current impl of `sort` does not handle nor test the special case of an empty view. 
Passing an empty view to sort should not fail. 
Currently, the code fails because if we look here: 

https://github.com/kokkos/kokkos/blob/3f602b6f283f82ecee2648f5606214a633d5e737/algorithms/src/Kokkos_Sort.hpp#L607-L640

when `view` is empty, the reduction does not happen and so `result` contains the reduction_identity values. 
Since those values are *not* equal, the following if is *not* satisfied:
```cpp
if (result.min_val == result.max_val) return;
```

and the code fails because it hits this:
```cpp
if (std::is_floating_point<typename ViewType::non_const_value_type>::value) {
    KOKKOS_ASSERT(std::isfinite(static_cast<double>(result.max_val) -
                                static_cast<double>(result.min_val)));
  }
```

## `BinSort`

For BinSort there is a similar problem here:

https://github.com/kokkos/kokkos/blob/3f602b6f283f82ecee2648f5606214a633d5e737/algorithms/src/Kokkos_Sort.hpp#L327-L331

## Comments

Note that an empty view is not a precondition violation. 
Also, the std lib behaves similarly: if we try to call std::sort on an empty vector the code does not crash. 

When merging, need to be careful with #6081

### CODE REVIEW

**Review by masterleinad — COMMENTED** [review]
What about `BinSort::sort`?

**Comment by fnrizzi** [comment]
> What about `BinSort::sort`?

yes that too, i am working on it... have not pushed yet

**Comment by fnrizzi** [comment]
retest this please

File: algorithms/unit_tests/TestSort.hpp [hunk 1]
```diff
@@ -440,6 +440,30 @@ void test_sort_empty_view() {
...
+  Kokkos::BinSort<KeyViewType, BinOp_t> Sorter;
+
+  // does not matter if we use int or something else
+  Kokkos::View<int*, ExecutionSpace> v;
+
+  // test all exposed public sort methods
+  Sorter.sort(ExecutionSpace(), v, 3, 8);
+  ASSERT_TRUE(v.extent(0) == 0);
+
+  Sorter.sort(v, 3, 8);
```
Comment (dalg24) [inline algorithms/unit_tests/TestSort.hpp:462]: (Commenting here but this is for both occurrences)
In my opinion you are writing a precondition violation.  The range must be valid.  Try [0,0[

Comment (fnrizzi) [reply algorithms/unit_tests/TestSort.hpp:462]: I had to do this to make it work, becuse of PR6131

File: algorithms/src/Kokkos_Sort.hpp [hunk 2]
```diff
@@ -694,6 +711,9 @@ void sort(const Cuda& space,
 
 template <class ViewType>
 void sort(ViewType const& view) {
+  if (view.extent(0) == 0) {
+    return;
+  }
```
Comment (dalg24) [inline algorithms/src/Kokkos_Sort.hpp:716]: Nope I think you need to fence before you exit early,

File: algorithms/unit_tests/TestSort.hpp (hunk 1 above)
Comment (dalg24) [reply algorithms/unit_tests/TestSort.hpp]: Yes exactly

File: algorithms/src/Kokkos_Sort.hpp (hunk 2 above)
Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:716]: Why?

Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp:716]: Similar to deep_copy.  Overload that does not take an execution space is synchronous.

Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:718]: thanks for pointing this out, i think i fixed it. there was also another overload wihtout exespace agument to fix.

File: algorithms/src/Kokkos_Sort.hpp [hunk 3]
```diff
@@ -704,6 +726,10 @@ template <class ExecutionSpace, class ViewType>
 std::enable_if_t<Kokkos::is_execution_space<ExecutionSpace>::value> sort(
     const ExecutionSpace& exec, ViewType view, size_t const begin,
     size_t const end) {
```
Comment (dalg24) [inline algorithms/src/Kokkos_Sort.hpp:728]: (Not expecting you to resolve here)
Anyone knows why we provide that overload?
Why is it not sufficient to sort a subview?  Also why is it not deferring to the overload that does not take the begin and end range arguments?
@crtrott

File: algorithms/unit_tests/TestSort.hpp [hunk 4]
```diff
@@ -425,6 +425,43 @@ void test_sort(unsigned int N) {
   test_sort_integer_overflow<ExecutionSpace, unsigned long long>();
   test_sort_integer_overflow<ExecutionSpace, int>();
 }
+
+template <class ExecutionSpace>
+void test_sort_empty_view() {
+  // does not matter if we use int or something else
+  Kokkos::View<int*, ExecutionSpace> v;
```
Comment (dalg24) [inline algorithms/unit_tests/TestSort.hpp:432]:
```suggestion
  Kokkos::View<int*, ExecutionSpace> v("v", 0);
```

File: algorithms/unit_tests/TestSort.hpp [hunk 5]
```diff
@@ -425,6 +425,43 @@ void test_sort(unsigned int N) {
...
+
+  Kokkos::sort(v);
+  // nothing should have changed
+  ASSERT_TRUE(v.extent(0) == 0);
+
+  Kokkos::sort(ExecutionSpace(), v);
+  // nothing should have changed
+  ASSERT_TRUE(v.extent(0) == 0);
```
Comment (dalg24) [inline algorithms/unit_tests/TestSort.hpp:440]:
```suggestion
  // TODO check the synchronous behavior of the calls below
  ASSERT_NO_THROW(Kokkos::sort(ExecutionSpace(), v));
  ASSERT_NO_THROW(Kokkos::sort(v));

```

File: algorithms/src/Kokkos_Sort.hpp (hunk 3 above)
Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:728]: because this is calling the bin sort which has a method taking a "range" defined by indices.

Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:728]: and it works for dynamic views too. 
my plan is to revisit all this in a separate PR to make it more clear why we hae some things and what happens.

Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp:728]: My point is that if sorting a subview has the same effect, what is the point of this "non-standard" overload?

File: algorithms/src/Kokkos_Sort.hpp [hunk 6]
```diff
@@ -648,8 +656,11 @@ sort(const ExecutionSpace& exec,
 template <class DataType, class... Properties>
 void sort(const Experimental::SYCL& space,
           const Kokkos::View<DataType, Properties...>& view) {
-  using ViewType = Kokkos::View<DataType, Properties...>;
+  if (view.extent(0) == 0) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:661]: you mean this is not necessary because it just happens that oneapi handles that case anyway? 
I think it is important to not rely on assumptions about what happens impl-wise. 
basically, what i am saying is that for constraints and preconditons and corner cases, each overload of a function should be see at a high-level as:
```cpp
// constraints 
void function(...){. //overload 1
 // preconditions 
 // corner cases
 // impl call 
}

// constraints 
void function(...){. //overload 2
 // preconditions 
 // corner cases
 // impl call 
}
```
and so just because overload2 happens to call overload1 or something else, the preconditions or corner cases should be explicitly done in each. Suppose that we did not do this but just rely on the fact that the corner case is handled in overload1. Then someone later changes the impl and forgets about the corner case. This messes everything up.

[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:661]: @dalg24 makes sense?

File: algorithms/src/Kokkos_Sort.hpp (hunk 3 above)
Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:728]: because you need to have the begin/end info to access the correct permutation elements that are stored inside binsort.

File: algorithms/unit_tests/TestSort.hpp (hunk 4 above)
Comment (fnrizzi) [reply algorithms/unit_tests/TestSort.hpp:432]: this slipped , thought i did it too. thanks !

File: algorithms/src/Kokkos_Sort.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp:661]: I would be comfortable with deferring to Thrust/oneDPL/the standard C++ library for something that trivial.
It is less code to maintain on our side.  That said, your approach is fine.

File: algorithms/src/Kokkos_Sort.hpp (hunk 3 above)
Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp:728]: This makes sense in the context of BinSort but not for the Kokkos::sort algorithm.

Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:728]: i totally agree with you. In fact, this seems like this part of the sort api was designed with the assumption that underneath was going to call bin sort.

File: algorithms/unit_tests/TestSort.hpp [hunk 7]
```diff
@@ -425,6 +425,39 @@ void test_sort(unsigned int N) {
...
+void test_binsort_empty_view() {
+  // the bounds and extents used below are totally arbitrary
+  // and, in theory, should have no impact
+
+  using KeyViewType = Kokkos::View<int*, ExecutionSpace>;
+  KeyViewType kv("kv", 20);
+
+  using BinOp_t = Kokkos::BinOp1D<KeyViewType>;
+  BinOp_t binOp(5, 0, 10);
+  Kokkos::BinSort<KeyViewType, BinOp_t> Sorter(ExecutionSpace{}, kv, binOp);
```
Comment (masterleinad) [inline algorithms/unit_tests/TestSort.hpp:449]: Maybe also test that `kv` can be empty?

Comment (fnrizzi) [reply algorithms/unit_tests/TestSort.hpp:449]: added another test

File: algorithms/src/Kokkos_Sort.hpp (hunk 3 above)
Comment (crtrott) [reply algorithms/src/Kokkos_Sort.hpp:728]: Yeah lets revisit, but not here.

Comment (fnrizzi) [reply algorithms/src/Kokkos_Sort.hpp:728]: yes not here, i will work on a new PR (I already started)

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
What did you change?

**Comment by fnrizzi** [comment]
> What did you change?

change what? I only rebased on develop

**Comment by dalg24** [comment]
> > What did you change?
> 
> change what? I only rebased on develop

Why bother?  Anyway too late.  I plan on merging after the CI completes

**Comment by fnrizzi** [comment]
> > > What did you change?
> > 
> > 
> > change what? I only rebased on develop
> 
> Why bother? Anyway too late. I plan on merging after the CI completes

habit of keeping things updated so that no surprises come up *after* merging (maybe in this case does not matter, but in general does)

**Comment by dalg24** [comment]
So you know we test your PR merged into develop unless there is conflict in which we can't merge anyway.

**Comment by fnrizzi** [comment]
> So you know we test your PR merged into develop unless there is conflict in which we can't merge anyway.

Now I know :) thanks!
